File: src/main/java/com/andrew121410/mc/ccminecraftbot/world/chunks/ChunkCache.java
package com.andrew121410.mc.ccminecraftbot.world.chunks;

import com.andrew121410.mc.ccminecraftbot.CCBotMinecraft;
import com.andrew121410.mc.ccminecraftbot.objects.Block;
import com.andrew121410.mc.ccminecraftbot.objects.BoundingBox;
import com.andrew121410.mc.ccminecraftbot.utils.ResourceManager;
import com.andrew121410.mc.ccminecraftbot.world.Location;
import com.github.steveice10.mc.protocol.data.game.chunk.Chunk;
import com.github.steveice10.mc.protocol.data.game.chunk.Column;
import com.github.steveice10.mc.protocol.data.game.entity.metadata.Position;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import java.util.HashMap;
import java.util.Map;

@Getter
@ToString
@EqualsAndHashCode
public class ChunkCache {

    private Map<ChunkPosition, Column> chunks;

    private CCBotMinecraft CCBotMinecraft;

    public ChunkCache(CCBotMinecraft CCBotMinecraft) {
        this.CCBotMinecraft = CCBotMinecraft;
        this.chunks = new HashMap<>();
    }

    public void addChunk(Column column) {
        ChunkPosition position = new ChunkPosition(column.getX(), column.getZ());
        this.chunks.put(position, column);
    }

    public void removeChunk(ChunkPosition chunkPosition) {
        this.chunks.remove(chunkPosition);
    }

    public void updateBlock(Position position, int block) {
        ChunkPosition chunkPosition = new ChunkPosition(position.getX() >> 4, position.getZ() >> 4);
        if (!chunks.containsKey(chunkPosition)) {
            return;
        }
        Column column = chunks.get(chunkPosition);
        Chunk chunk = column.getChunks()[position.getY() >> 4];
        Position blockPosition = chunkPosition.getChunkBlock(position.getX(), position.getY(), position.getZ());
        chunk.set(blockPosition.getX(), blockPosition.getY(), blockPosition.getZ(), block);
    }

    public Integer getBlockID(Position position) {
        ChunkPosition chunkPosition = new ChunkPosition(position.getX() >> 4, position.getZ() >> 4);
        if (!chunks.containsKey(chunkPosition)) {
            return null;
        }
        Column column = chunks.get(chunkPosition);
        Chunk chunk = column.getChunks()[position.getY() >> 4];
        Position blockPosition = chunkPosition.getChunkBlock(position.getX(), position.getY(), position.getZ());
        return chunk.get(blockPosition.getX(), blockPosition.getY(), blockPosition.getZ());
    }

    public Block getBlock(Position position) {
        return ResourceManager.INSTANCE.getBlocks().get(getBlockID(position));
    }

    public Block getBlock(Location location) {
        return ResourceManager.INSTANCE.getBlocks().get(getBlockID(location.toPosition()));
    }

    public boolean isSolid(Position position) {
        return getBlock(position).getBoundingBox() == BoundingBox.block;
    }

    public boolean isSolid(Location position) {
        return getBlock(position).getBoundingBox() == BoundingBox.block;
    }
}

File: src/main/kotlin/com/andrew121410/mc/ccminecraftbot/world/chunks/ChunkCache.kt
package com.andrew121410.mc.ccminecraftbot.world.chunks

import com.andrew121410.mc.ccminecraftbot.objects.Block
import com.andrew121410.mc.ccminecraftbot.objects.BoundingBox
import com.andrew121410.mc.ccminecraftbot.utils.ResourceManager
import com.andrew121410.mc.ccminecraftbot.world.Location
import com.github.steveice10.mc.protocol.data.game.chunk.Chunk
import com.github.steveice10.mc.protocol.data.game.chunk.Column
import com.github.steveice10.mc.protocol.data.game.entity.metadata.Position

data class ChunkCache(private val chunks: MutableMap<ChunkPosition, Column> = HashMap()) {
    fun addToCache(chunk: Column): Column? {
        val chunkPosition = ChunkPosition(chunk.x, chunk.z)
        val existingChunk: Column
        if (chunk.biomeData == null) return null

        if (chunks.containsKey(chunkPosition)) { // Column is already present in cache, we can merge with existing\
            existingChunk = chunks[chunkPosition]!!
            var changed = false
            for (i in chunk.chunks.indices) { // The chunks member is final, so chunk.getChunks() will probably be inlined and then completely optimized away
                if (chunk.chunks[i] != null) {
                    existingChunk.chunks[i] = chunk.chunks[i]
                    changed = true
                }
            }
            if (changed) existingChunk else null
        } else {
            chunks[chunkPosition] = chunk
            chunk
        }
        return null
    }

    fun removeChunk(chunkX: Int, chunkZ: Int) {
        val chunkPosition = ChunkPosition(chunkX, chunkZ)
        chunks.remove(chunkPosition)
    }

    fun updateBlock(location: Location, block: Int) {
        val x = location.x.toInt()
        val y = location.y.toInt()
        val z = location.z.toInt()

        val column = getChunk(x shr 4, z shr 4) ?: return

        if (y < MINIMUM_WORLD_HEIGHT || y shr 4 > column.chunks.size - 1) {
            // Y likely goes above or below the height limit of this world
            return
        }

        val chunk = column.chunks[y shr 4] ?: return
        chunk[x and 0xF, y and 0xF, z and 0xF] = block
    }

    fun getChunk(chunkX: Int, chunkZ: Int): Column? {
        val chunkPosition = ChunkPosition(chunkX, chunkZ)
        return chunks.getOrDefault(chunkPosition, null)
    }


    fun getBlockID(location: Location): Int {
        val x = location.x.toInt()
        val y = location.y.toInt()
        val z = location.z.toInt()

        val column = getChunk(x shr 4, z shr 4) ?: return 0
        if (y < MINIMUM_WORLD_HEIGHT || y shr 4 > column.chunks.size - 1) {
            // Y likely goes above or below the height limit of this world
            return 0
        }
        val chunk: Chunk = column.chunks[y shr 4] ?: return 0
        return chunk.get(x and 0xF, y and 0xF, z and 0xF) ?: return 0
    }

    fun getBlock(location: Location): Block {
        val blockID: Int = getBlockID(location)
        val block: Block = ResourceManager.getBlockByStateID(blockID)
        return block
    }

    fun isSolid(position: Position): Boolean {
        return getBlock(Location.from(position)).boundingBox === BoundingBox.block
    }

    fun isSolid(position: Location): Boolean {
        return getBlock(position).boundingBox === BoundingBox.block
    }

    companion object {
        private const val MINIMUM_WORLD_HEIGHT = 0
    }
}
